sklearn/preprocessing/_csr_polynomial_expansion.pyx

```diff
@@ -1,59 +1,174 @@
-# Author: Andrew nystrom <awnystrom@gmail.com>
+# Authors: Andrew nystrom <awnystrom@gmail.com>
+#          Meekail Zain <zainmeekail@gmail.com>
+from libc.limits cimport INT_MAX, LONG_MAX
+from libc.math cimport sqrt, pow
+cimport numpy as cnp
+cnp.import_array()
 
 from scipy.sparse import csr_matrix
-cimport numpy as cnp
+import numpy as np
 
-cnp.import_array()
-ctypedef cnp.int32_t INDEX_T
+ctypedef cnp.int8_t FLAG_t
+
+# This is a conservative upper bound for the maximum value before the
+# intermediate computation i**2 + 3*i overflows.
+# Corresponds to to i**2 + 3*i = maxint32
+cdef cnp.int64_t MAX_SAFE_INDEX_DEG2 = <cnp.int64_t> sqrt(LONG_MAX) - 4
+
+# Note that for the equations below, d=`n_features`
 
-ctypedef fused DATA_T:
+# This is a conservative upper bound for the maximum value before the
+# intermediate computation 3 * d**2 * d + d**3 overflows. We leverage the fact
```

Comment:
As discussed IRL, I think we need a comment that connects the equation in `_deg3_column` to this approximate bound.